# <center> Assignment 2
    
    
    Submitted by: DIBYA RANJAN DASH
    BRANCH: CEN
    SIC: 20BECB77

# Part A: Data Preprocessing

## 1. Import Libraries

In [34]:
import numpy as np
import tensorflow as tf
import pandas as pd
import matplotlib.pylab as plt

## 2. Import Dataset

In [35]:
dataset = pd.read_csv('pima-indians-diabetes.data.csv')

In [36]:
dataset

6  148  72  35    0  33.6  0.627  50  1
0     1   85  66  29    0  26.6  0.351  31  0
1     8  183  64   0    0  23.3  0.672  32  1
2     1   89  66  23   94  28.1  0.167  21  0
3     0  137  40  35  168  43.1  2.288  33  1
4     5  116  74   0    0  25.6  0.201  30  0
..   ..  ...  ..  ..  ...   ...    ...  .. ..
762  10  101  76  48  180  32.9  0.171  63  0
763   2  122  70  27    0  36.8  0.340  27  0
764   5  121  72  23  112  26.2  0.245  30  0
765   1  126  60   0    0  30.1  0.349  47  1
766   1   93  70  31    0  30.4  0.315  23  0

[767 rows x 9 columns]

## 3. Create feature matrix X and dependent variable Y

In [37]:
X = dataset.iloc[:,0:8].values
Y = dataset.iloc[:,8].values

## 4. Splitting Training and Testing data

In [38]:
from sklearn.model_selection import train_test_split
Xtrain,Xtest,Ytrain,Ytest = train_test_split(X,Y,test_size = 0.2,random_state=1)

In [39]:
Xtrain

array([[9.00e+00, 1.45e+02, 8.00e+01, ..., 3.79e+01, 6.37e-01, 4.00e+01],
       [1.00e+01, 1.29e+02, 6.20e+01, ..., 4.12e+01, 4.41e-01, 3.80e+01],
       [0.00e+00, 1.14e+02, 8.00e+01, ..., 4.42e+01, 1.67e-01, 2.70e+01],
       ...,
       [4.00e+00, 1.29e+02, 8.60e+01, ..., 3.51e+01, 2.31e-01, 2.30e+01],
       [7.00e+00, 1.81e+02, 8.40e+01, ..., 3.59e+01, 5.86e-01, 5.10e+01],
       [2.00e+00, 9.00e+01, 6.80e+01, ..., 3.82e+01, 5.03e-01, 2.70e+01]])

## 5. Feature Scaling

In [40]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
Xtrain[:,:] = sc.fit_transform(Xtrain[:,:])
Xtest[:,:] = sc.fit_transform(Xtest[:,:])

# Part B: Building Model

## a) Building Decision Tree Classifier

### Training Model

In [41]:
from sklearn.tree import DecisionTreeClassifier
DT = DecisionTreeClassifier(criterion='entropy', random_state = 0)
DT.fit(Xtrain,Ytrain)

DecisionTreeClassifier(criterion='entropy', random_state=0)

### Testing Model

In [42]:
Yestimated = DT.predict(Xtest)
print(np.concatenate((Yestimated.reshape(len(Yestimated),1),Ytest.reshape(len(Ytest),1)),1))

[[0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 0]
 [0 0]
 [1 1]
 [1 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [1 0]
 [0 0]
 [0 0]
 [0 0]
 [0 1]
 [1 0]
 [0 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 0]
 [1 1]
 [1 0]
 [0 1]
 [0 0]
 [0 0]
 [0 1]
 [0 0]
 [0 1]
 [0 0]
 [0 1]
 [1 1]
 [0 0]
 [0 0]
 [0 1]
 [0 1]
 [0 0]
 [0 0]
 [0 0]
 [1 0]
 [0 0]
 [1 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [1 0]
 [1 0]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [1 0]
 [1 0]
 [1 0]
 [0 0]
 [1 0]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 0]
 [1 1]
 [1 1]
 [0 1]
 [1 0]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [0 1]
 [1 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 0]
 [1 0]
 [0 0]
 [1 0]
 [1 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [1 0]
 [1 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [1 0]
 [0 0]
 [1 1]
 [1 0]
 [0 1]
 [0 0]
 [1 1]
 [0 1]
 [0 0]
 [0 0]
 [1 0]
 [1 0]
 [1 1]
 [1 0]
 [1 0]
 [1 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 1]
 [0 0]

### Performance Metric of Decision Tree Classifier

In [59]:
from sklearn.metrics import confusion_matrix,accuracy_score,precision_score,mean_squared_error
cm = confusion_matrix(Ytest , Yestimated)
print(f' Confusion Matrix: \n {cm}')
print(f'\n accuracy_score: {accuracy_score(Ytest , Yestimated)} ')
print(f'\n precision_score: {precision_score(Ytest , Yestimated)} ')
print(f'\n mean square error: {mean_squared_error(Ytest,Yestimated)}')
print(f'\n root mean square error: {np.sqrt(mean_squared_error(Ytest,Yestimated))}')

 Confusion Matrix: 
 [[88 20]
 [13 33]]

 accuracy_score: 0.7857142857142857 

 precision_score: 0.6226415094339622 

 mean square error: 0.21428571428571427

 root mean square error: 0.4629100498862757


## b) Random Forest Classifier

### Training the model

In [44]:
from sklearn.ensemble import RandomForestClassifier
rfc = RandomForestClassifier(n_estimators=40,criterion='entropy',random_state=0)
rfc.fit(Xtrain,Ytrain)

RandomForestClassifier(criterion='entropy', n_estimators=40, random_state=0)

### Testing model

In [45]:
Yestimated = rfc.predict(Xtest)
Yestimated

array([1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0,
       0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1,
       1, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0],
      dtype=int64)

### Performance Metric of Decision Tree Classifier

In [60]:
from sklearn.metrics import confusion_matrix,accuracy_score,precision_score
cm = confusion_matrix(Ytest , Yestimated)
print(f' Confusion Matrix: \n {cm}')
print(f'\n accuracy_score: {accuracy_score(Ytest , Yestimated)} ')
print(f'\n precision_score: {precision_score(Ytest , Yestimated)} ')
print(f'\n mean square error: {mean_squared_error(Ytest,Yestimated)}')
print(f'\n root mean square error: {np.sqrt(mean_squared_error(Ytest,Yestimated))}')

 Confusion Matrix: 
 [[88 20]
 [13 33]]

 accuracy_score: 0.7857142857142857 

 precision_score: 0.6226415094339622 

 mean square error: 0.21428571428571427

 root mean square error: 0.4629100498862757


## c) ANN_Model

In [47]:
d1 = tf.keras.models.Sequential()

In [48]:
# create model

# First Hidden Layer
d1.add(tf.keras.layers.Dense(units=12,activation='relu'))

# Second Hidden Layer
d1.add(tf.keras.layers.Dense(units=8,activation='relu'))

# Output Layer
d1.add(tf.keras.layers.Dense(units=1,activation='sigmoid'))

### Compile the Model

In [49]:
d1.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

### Training the model

In [50]:
d1.fit(Xtrain,Ytrain)

20/20 [==============================] - 1s 3ms/step - loss: 0.7100 - accuracy: 0.6150


In [51]:
d1.fit(Xtrain,Ytrain,epochs=200)

Epoch 1/200
20/20 [==============================] - 0s 3ms/step - loss: 0.6746 - accuracy: 0.6444
Epoch 2/200
20/20 [==============================] - 0s 3ms/step - loss: 0.6462 - accuracy: 0.6623
Epoch 3/200
20/20 [==============================] - 0s 3ms/step - loss: 0.6226 - accuracy: 0.6672
Epoch 4/200
20/20 [==============================] - 0s 3ms/step - loss: 0.5995 - accuracy: 0.6803
Epoch 5/200
20/20 [==============================] - 0s 3ms/step - loss: 0.5810 - accuracy: 0.6917
Epoch 6/200
20/20 [==============================] - 0s 3ms/step - loss: 0.5644 - accuracy: 0.6998
Epoch 7/200
20/20 [==============================] - 0s 3ms/step - loss: 0.5508 - accuracy: 0.7047
Epoch 8/200
20/20 [==============================] - 0s 2ms/step - loss: 0.5417 - accuracy: 0.7227
Epoch 9/200
20/20 [==============================] - 0s 2ms/step - loss: 0.5319 - accuracy: 0.7276
Epoch 10/200
20/20 [==============================] - 0s 2ms/step - loss: 0.5245 - accuracy: 0.7325
Epoch 11/

20/20 [==============================] - 0s 2ms/step - loss: 0.4297 - accuracy: 0.7961
Epoch 84/200
20/20 [==============================] - 0s 1ms/step - loss: 0.4293 - accuracy: 0.7928
Epoch 85/200
20/20 [==============================] - 0s 2ms/step - loss: 0.4284 - accuracy: 0.8010
Epoch 86/200
20/20 [==============================] - 0s 2ms/step - loss: 0.4282 - accuracy: 0.7961
Epoch 87/200
20/20 [==============================] - 0s 2ms/step - loss: 0.4280 - accuracy: 0.8010
Epoch 88/200
20/20 [==============================] - 0s 2ms/step - loss: 0.4273 - accuracy: 0.8010
Epoch 89/200
20/20 [==============================] - 0s 2ms/step - loss: 0.4265 - accuracy: 0.8075
Epoch 90/200
20/20 [==============================] - 0s 2ms/step - loss: 0.4263 - accuracy: 0.8091
Epoch 91/200
20/20 [==============================] - 0s 2ms/step - loss: 0.4251 - accuracy: 0.8042
Epoch 92/200
20/20 [==============================] - 0s 2ms/step - loss: 0.4252 - accuracy: 0.8026
Epoch 93/200


20/20 [==============================] - 0s 2ms/step - loss: 0.3907 - accuracy: 0.8075
Epoch 165/200
20/20 [==============================] - 0s 2ms/step - loss: 0.3903 - accuracy: 0.8157
Epoch 166/200
20/20 [==============================] - 0s 2ms/step - loss: 0.3900 - accuracy: 0.8124
Epoch 167/200
20/20 [==============================] - 0s 2ms/step - loss: 0.3898 - accuracy: 0.8091
Epoch 168/200
20/20 [==============================] - 0s 2ms/step - loss: 0.3891 - accuracy: 0.8108
Epoch 169/200
20/20 [==============================] - 0s 2ms/step - loss: 0.3883 - accuracy: 0.8108
Epoch 170/200
20/20 [==============================] - 0s 2ms/step - loss: 0.3883 - accuracy: 0.8091
Epoch 171/200
20/20 [==============================] - 0s 2ms/step - loss: 0.3874 - accuracy: 0.8091
Epoch 172/200
20/20 [==============================] - 0s 2ms/step - loss: 0.3883 - accuracy: 0.8075
Epoch 173/200
20/20 [==============================] - 0s 2ms/step - loss: 0.3867 - accuracy: 0.8108
Epoc

### Testing the model

In [52]:
Yestimated = d1.predict(Xtest)

5/5 [==============================] - 0s 0s/step


In [53]:
Ytest

array([0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1,
       0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1,
       0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1,
       1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0,
       1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0],
      dtype=int64)

In [54]:
Yestimated

array([[0.566197  ],
       [0.05038987],
       [0.08172856],
       [0.02051053],
       [0.06823369],
       [0.6307413 ],
       [0.01680713],
       [0.5751395 ],
       [0.20953341],
       [0.00104568],
       [0.63628143],
       [0.01140257],
       [0.04422596],
       [0.10045037],
       [0.05293845],
       [0.65643394],
       [0.44504148],
       [0.37968788],
       [0.19275272],
       [0.12663403],
       [0.0013706 ],
       [0.58794314],
       [0.65434563],
       [0.30694845],
       [0.18795078],
       [0.03226232],
       [0.15168604],
       [0.01736138],
       [0.08745633],
       [0.82874376],
       [0.90886825],
       [0.32255307],
       [0.40026933],
       [0.56781775],
       [0.02229736],
       [0.8465591 ],
       [0.04647144],
       [0.37181103],
       [0.90101343],
       [0.30254924],
       [0.5544233 ],
       [0.28645545],
       [0.40682104],
       [0.44077685],
       [0.01153498],
       [0.6039191 ],
       [0.03847978],
       [0.141

In [55]:
Yestimated = (Yestimated>0.5)

In [56]:
Yestimated

array([[ True],
       [False],
       [False],
       [False],
       [False],
       [ True],
       [False],
       [ True],
       [False],
       [False],
       [ True],
       [False],
       [False],
       [False],
       [False],
       [ True],
       [False],
       [False],
       [False],
       [False],
       [False],
       [ True],
       [ True],
       [False],
       [False],
       [False],
       [False],
       [False],
       [False],
       [ True],
       [ True],
       [False],
       [False],
       [ True],
       [False],
       [ True],
       [False],
       [False],
       [ True],
       [False],
       [ True],
       [False],
       [False],
       [False],
       [False],
       [ True],
       [False],
       [False],
       [False],
       [False],
       [False],
       [ True],
       [False],
       [ True],
       [False],
       [False],
       [False],
       [ True],
       [ True],
       [ True],
       [False],
       [ True],
       [

### Performance Metric

In [57]:
from sklearn.metrics import confusion_matrix,accuracy_score,precision_score
cm = confusion_matrix(Ytest , Yestimated)
print(f' Confusion Matrix: \n {cm}')
print(f'\n accuracy_score: {accuracy_score(Ytest , Yestimated)} ')
print(f'\n precision_score: {precision_score(Ytest , Yestimated)} ')
print(f'\n mean square error: {mean_squared_error(Ytest,Yestimated)}')
print(f'\n root mean square error: {np.sqrt(mean_squared_error(Ytest,Yestimated))}')

 Confusion Matrix: 
 [[88 20]
 [13 33]]

 accuracy_score: 0.7857142857142857 

 precision_score: 0.6226415094339622 

 mean square error: 0.21428571428571427

 root mean square error: 0.4629100498862757


### So the Accuracy of all the models are as folllows: 
 >a) Decision Tree Classifier Model: 0.70

 >b) Random Forest Classifier Model: 0.75
 
 >c) ANN Model: 0.77

Here the Accuracy score of __Decision Tree__ is __70%__ , __Random Forest__ is __75%__ and __ANN Model__ is __77%__  for the __pima-indians-diabetes.data.csv__ Dataset. Among them the __ANN Model__ got the highest accuracy score

-- Hence in this case __ANN Model__ is the best one